src/lib.rs

```diff
@@ -1780,4 +1801,96 @@ mod test {
+        let mut mock_gateway = MockServerGatewayApis::new();
+        mock_gateway
+            .expect_complete_activity_task()
+            .times(1)
+            .returning(|_, _| Err(tonic::Status::not_found("unimportant")));
+
+        let core = CoreSDK::new(
+            mock_gateway,
+            CoreInitOptions {
+                gateway_opts: fake_sg_opts(),
+                evict_after_pending_cleared: true,
+                max_outstanding_workflow_tasks: 5,
+                max_outstanding_activities: 5,
+            },
+        );
+
+        core.complete_activity_task(ActivityTaskCompletion {
+            task_token: vec![1],
+            result: Some(ActivityResult::ok(vec![1].into())),
+        })
+        .await
+        .unwrap();
+    }
+
+    #[tokio::test]
+    async fn activity_cancel_interrupts_poll() {
+        let mut mock_gateway = MockManualGateway::new();
+        let mut poll_resps = VecDeque::from(vec![
+            async {
+                Ok(PollActivityTaskQueueResponse {
+                    task_token: vec![1],
+                    heartbeat_timeout: Some(Duration::from_secs(1).into()),
+                    ..Default::default()
+                })
+            }
+            .boxed(),
+            async {
+                tokio::time::sleep(Duration::from_millis(500)).await;
+                Ok(Default::default())
+            }
+            .boxed(),
+        ]);
+        mock_gateway
+            .expect_poll_activity_task()
+            .times(2)
+            .returning(move || poll_resps.pop_front().unwrap());
+        mock_gateway
+            .expect_record_activity_heartbeat()
+            .times(1)
+            .returning(|_, _| {
+                async {
+                    Ok(RecordActivityTaskHeartbeatResponse {
+                        cancel_requested: true,
+                    })
+                }
+                .boxed()
+            });
+
+        let core = CoreSDK::new(
+            mock_gateway,
+            CoreInitOptions {
+                gateway_opts: fake_sg_opts(),
+                evict_after_pending_cleared: true,
+                max_outstanding_workflow_tasks: 5,
+                max_outstanding_activities: 5,
+            },
+        );
+        let last_finisher = AtomicUsize::new(0);
+        // Perform first poll to get the activity registered
+        let act = core.poll_activity_task().await.unwrap();
+        // Poll should block until heartbeat is sent, issuing the cancel, and interrupting the poll
+        tokio::join! {
+            async {
+                core.record_activity_heartbeat(ActivityHeartbeat {
+                    task_token: act.task_token,
+                    details: vec![vec![1u8, 2, 3].into()],
+                })
+                .await
+                .unwrap();
+                last_finisher.store(1, Ordering::SeqCst);
+            },
+            async {
+                core.poll_activity_task().await.unwrap();
```

Comment:
I feel like we're missing an explicit assertion here that poll returns `ActivityTask.Cancel`.